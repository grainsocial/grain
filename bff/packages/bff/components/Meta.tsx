import type { JSX } from "preact";

export type MetaDescriptor =
  | { charSet: "utf-8" }
  | { title: string }
  | { name: string; content: string }
  | { property: string; content: string }
  | { tagName: "meta" | "link"; [name: string]: string }
  | { [name: string]: unknown };

/**
 * Meta component for setting meta tags in the document head.
 *
 * @example
 * ```tsx
 * <Meta
 *   meta={[
 *     { charset: "utf-8" },
 *     { title: "My Page Title" },
 *     { name: "description", content: "My page description" },
 *     { property: "og:title", content: "My Page Title" },
 *     { property: "og:description", content: "My page description" },
 *     { property: "og:image", content: "https://example.com/image.jpg" },
 *     { property: "og:url", content: "https://example.com" },
 *     { tagName: "link", rel: "icon", href: "/favicon.ico" },
 *     { tagName: "link", rel: "stylesheet", href: "/styles.css" },
 *     { tagName: "meta", name: "viewport", content: "width=device-width, initial-scale=1" },
 *   ]}
 * />
 * ```
 *
 * @param meta - An array of meta descriptors.
 * @returns A JSX element representing the meta tags.
 */
export function Meta(
  { meta = [] }: Readonly<{ meta?: MetaDescriptor[] }>,
): JSX.Element {
  return (
    <>
      {meta.map((metaProps) => {
        if (!metaProps) {
          return null;
        }

        if ("title" in metaProps) {
          return <title key="title">{String(metaProps.title)}</title>;
        }

        if ("charset" in metaProps) {
          metaProps.charSet ??= metaProps.charset;
          delete metaProps.charset;
        }

        if ("charSet" in metaProps && metaProps.charSet != null) {
          return typeof metaProps.charSet === "string"
            ? <meta key="charSet" charSet={metaProps.charSet} />
            : null;
        }

        if ("tagName" in metaProps) {
          const { tagName, ...rest } = metaProps;
          if (!isValidMetaTag(tagName)) {
            console.warn(
              `A meta object uses an invalid tagName: ${tagName}. Expected either 'link' or 'meta'`,
            );
            return null;
          }
          const Comp = tagName;
          return <Comp key={JSON.stringify(rest)} {...rest} />;
        }

        return <meta key={JSON.stringify(metaProps)} {...metaProps} />;
      })}
    </>
  );
}

function isValidMetaTag(tagName: unknown): tagName is "meta" | "link" {
  return typeof tagName === "string" && /^(meta|link)$/.test(tagName);
}
