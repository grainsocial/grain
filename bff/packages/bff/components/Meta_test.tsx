import { assertSnapshot } from "@std/testing/snapshot";
import { renderToString } from "preact-render-to-string";
import { Meta } from "./Meta.tsx";
import type { MetaDescriptor } from "./mod.tsx";

const render = (meta: MetaDescriptor[]) => {
  return renderToString(<Meta meta={meta} />);
};

Deno.test("meta with charSet", async (t) => {
  const result = render([{ charSet: "utf-8" }]);
  await assertSnapshot(t, result);
});

Deno.test("meta with title", async (t) => {
  const result = render([{ title: "Test Title" }]);
  await assertSnapshot(t, result);
});

Deno.test("meta with name and content", async (t) => {
  const result = render([
    { name: "description", content: "Test description" },
  ]);
  await assertSnapshot(t, result);
});

Deno.test("meta with property and content", async (t) => {
  const result = render([
    { property: "og:title", content: "Test Open Graph Title" },
  ]);
  await assertSnapshot(t, result);
});

Deno.test("meta with tagName", async (t) => {
  const result = render([
    { tagName: "link", rel: "stylesheet", href: "/styles.css" },
  ]);
  await assertSnapshot(t, result);
});

Deno.test("meta with invalid tagName", async (t) => {
  const result = render([
    { tagName: "invalid", rel: "stylesheet", href: "/styles.css" },
  ]);
  await assertSnapshot(t, result);
});

Deno.test("meta with multiple descriptors", async (t) => {
  const result = render([
    { charSet: "utf-8" },
    { title: "Test Title" },
    { name: "description", content: "Test description" },
    { property: "og:title", content: "Test Open Graph Title" },
    { tagName: "link", rel: "stylesheet", href: "/styles.css" },
  ]);
  await assertSnapshot(t, result);
});
