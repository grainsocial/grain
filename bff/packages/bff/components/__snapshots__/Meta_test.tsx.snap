export const snapshot = {};

snapshot[`meta with charSet 1`] = `'<meta  charset="utf-8">'`;

snapshot[`meta with title 1`] = `"<title >Test Title</title>"`;

snapshot[`meta with name and content 1`] = `'<meta name="description" content="Test description"/>'`;

snapshot[`meta with property and content 1`] = `'<meta property="og:title" content="Test Open Graph Title"/>'`;

snapshot[`meta with tagName 1`] = `'<link rel="stylesheet" href="/styles.css"/>'`;

snapshot[`meta with invalid tagName 1`] = `""`;

snapshot[`meta with multiple descriptors 1`] = `'<meta  charset="utf-8"><title >Test Title</title><meta name="description" content="Test description"/><meta property="og:title" content="Test Open Graph Title"/><link rel="stylesheet" href="/styles.css"/>'`;
